broker-core/src/main/java/io/zeebe/broker/subscription/message/state/MessageStateController.java

```diff
@@ -185,7 +195,37 @@ private int writeMessageIdKey(
     return offset;
   }
 
-  public Message findFirstMessage(final DirectBuffer name, final DirectBuffer correlationKey) {
+  private int writeCorrelatedMessageKey(
+      MutableDirectBuffer buffer, long messageKey, long workflowInstanceKey) {
+    int offset = 0;
+
+    buffer.putLong(offset, messageKey, STATE_BYTE_ORDER);
+    offset += Long.BYTES;
+
+    buffer.putLong(offset, workflowInstanceKey, STATE_BYTE_ORDER);
+    offset += Long.BYTES;
+    return offset;
+  }
+
+  public void putMessageCorrelation(long messageKey, long workflowInstanceKey) {
+    final int length = writeCorrelatedMessageKey(keyBuffer, messageKey, workflowInstanceKey);
+    db.put(
+        correlatedMessageColumnFamily,
+        keyBuffer.byteArray(),
+        0,
+        length,
+        EXISTENCE,
+        0,
+        EXISTENCE.length);
+  }
+
+  public boolean existMessageCorrelation(long messageKey, long workflowInstanceKey) {
+    final int length = writeCorrelatedMessageKey(keyBuffer, messageKey, workflowInstanceKey);
+    return db.exists(correlatedMessageColumnFamily, keyBuffer.byteArray(), 0, length);
+  }
+
+  public void findMessage(
```

Comment:
I think the naming is not clear to what the method really does

Reply:
What about `visitMessages(...)`?